watertap/core/zero_order_siso.py

```diff
@@ -0,0 +1,216 @@
+"""
+This module contains the methods for constructing the material balances for
+zero-order single-input/single-output (SISO) unit models (i.e. units with a single inlet and single
+outlet where composition changes, such as a generic bioreactor).
+"""
+
+import idaes.logger as idaeslog
+from idaes.core.util import get_solver
+import idaes.core.util.scaling as iscale
+
+from pyomo.environ import NonNegativeReals, Var, units as pyunits
+
+# Some more information about this module
+__author__ = "Adam Atia"
+
+# Set up logger
+_log = idaeslog.getLogger(__name__)
+
+
+def build_siso(self):
+    """
+    Helper method for constructing material balances for zero-order type models
+    with SISO behavior.
+
+    Two StateBlocks are added with two corresponding Ports:
+        * properties_in --> inlet
+        * properties_treated ---> treated
+
+#todo: revisit this after formulation complete
+
+    One additional variable is added:
+        * removal_frac_mass_solute (indexed by time and solute)
+
+    Two additional constraints are added to represent the material balances
+        * flow_balance (indexed by time)
+        * solute_treated_equation (indexed by time and solute)
+
+    This method also sets private attributes on the unit model with references
+    to the appropriate initialization and scaling methods to use and to return
+    the inlet volumetric flow rate.
+    """
+    self._has_recovery_removal = True
+    self._initialize = initialize_siso
+    self._scaling = calculate_scaling_factors_siso
+
+    # Create state blocks for inlet and outlets
+    tmp_dict = dict(**self.config.property_package_args)
+    tmp_dict["has_phase_equilibrium"] = False
+    tmp_dict["defined_state"] = True
+
+    self.properties_in = self.config.property_package.build_state_block(
+        self.flowsheet().time,
+        doc="Material properties at inlet",
+        default=tmp_dict)
+
+    tmp_dict_2 = dict(**tmp_dict)
+    tmp_dict_2["defined_state"] = False
+
+    self.properties_treated = \
+        self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of treated water",
+            default=tmp_dict_2)
+
+    # Create Ports
+    self.add_port("inlet", self.properties_in, doc="Inlet port")
+    self.add_port("treated",
+                  self.properties_treated,
+                  doc="Treated water outlet port")
+
+    # Add performance variables
+    self.removal_frac_mass_solute = Var(
+        self.flowsheet().time,
+        self.config.property_package.solute_set,
+        domain=NonNegativeReals,
+        initialize=0.01,
+        units=pyunits.dimensionless,
+        doc='Solute removal fraction on a mass basis')
+
+    # Add performance constraints
+
+    # Flow balance
+    @self.Constraint(self.flowsheet().time, doc='Overall flow balance')
+    def water_balance(b, t):
+        return (b.properties_in[t].flow_mass_comp["H2O"] ==
```

Comment:
I know I said otherwise at the beginning, but seeing as we are already throwing conservation of mass out the window with this for solutes, should we extent the same approach to water? I.e. include a recovery and just assume it is 1 unless we are told otherwise?

The YAML files could still leave out recovery in most cases, and only add it in for those where is is not 1.

I could see this as maybe being useful for evaporation ponds, where water leaves the system, but we do not capture the evaporated water.

Reply:
On that I am not sure - lets leave it as is for now.